### PR TITLE
doc(client, service): Update documents about the Proxy property for various clients

### PR DESCRIPTION
**Pull request by jamdavi**
In issue #2242 it was clear that we do not properly document the Proxy property. This will add additional details about this property and others for various clients.

## Device Client
image

## Service Client
image

image

## Registry Manager and Job Client
image

### CODE REVIEW

File: iothub/device/src/ITransportSettings.cs [hunk 1]
```diff
@@ -20,5 +22,41 @@ public interface ITransportSettings
         /// The time to wait for a receive operation.
         /// </summary>
         TimeSpan DefaultReceiveTimeout { get; }
+
+        /// <summary>
+        /// The web proxy that will be used to connect to IoT Hub using a web socket connection for AMQP or MQTT; or when using the HTTP protocol.
```
Comment (drwill-ms) [inline iothub/device/src/ITransportSettings.cs:27]:
```suggestion
        /// The web proxy that will be used to connect to IoT hub using a web socket connection for AMQP, MQTT, or when using the HTTP protocol.
```

File: iothub/device/src/ITransportSettings.cs [hunk 2]
```diff
@@ -20,5 +22,41 @@ public interface ITransportSettings
...
+        /// To set a proxy you must instantiate an instance of the <see cref="WebProxy"/> class--or any class that derives from <see cref="IWebProxy"/>. The snippet below shows a method that returns a device using a proxy that connects to localhost on port 8888.
+        /// <code>
+        /// static DeviceClient GetClientWithProxy()
+        /// {
+        ///     try
+        ///     {
+        ///         var proxyHost = "localhost";
+        ///         var proxyPort = 8888;
+        ///         // In the event the AMQP ports are blocked from this device, we will fall back to AMQP over web sockets
+        ///         var transportSettings = new AmqpTransportSettings(Microsoft.Azure.Devices.Client.TransportType.Amqp)
```
Comment (drwill-ms) [inline iothub/device/src/ITransportSettings.cs:45]: Shouldn't the best practice be that they specify Amqp over websocket explicitly?

File: iothub/service/src/HttpTransportSettings.cs [hunk 3]
```diff
@@ -20,12 +20,38 @@ public HttpTransportSettings()
         }
 
         /// <summary>
-        /// Proxy information.
+        /// The web proxy that will be used to connect to IoT Hub when using the HTTP protocol.
```
Comment (drwill-ms) [inline iothub/service/src/HttpTransportSettings.cs:23]:
```suggestion
        /// The web proxy that will be used to connect to IoT hub when using the HTTP protocol.
```

File: iothub/service/src/HttpTransportSettings.cs [hunk 4]
```diff
@@ -20,12 +20,38 @@ public HttpTransportSettings()
...
+        /// <example>
+        /// To set a proxy you must instantiate an instance of the <see cref="WebProxy"/> class--or any class that derives from <see cref="IWebProxy"/>. The snippet below shows a method that returns a device using a proxy that connects to localhost on port 8888.
+        /// <code>
+        /// static JobClient GetJobClient()
+        /// {
+        ///     try
+        ///     {
+        ///         var proxyHost = "localhost";
+        ///         var proxyPort = 8888;
+        ///         var transportSettings = new HttpTransportSettings()
```
Comment (drwill-ms) [inline iothub/service/src/HttpTransportSettings.cs:40]:
```suggestion
        ///         var transportSettings = new HttpTransportSettings
```

File: iothub/service/src/ServiceClientTransportSettings.cs [hunk 5]
```diff
@@ -21,13 +21,77 @@ public ServiceClientTransportSettings()
         }
 
         /// <summary>
-        /// The proxy settings to be used on the AMQP client.
+        /// The web proxy that will be used to connect to IoT Hub when using the AMQP over web sockets.
```
Comment (drwill-ms) [inline iothub/service/src/ServiceClientTransportSettings.cs:24]:
```suggestion
        /// The web proxy that will be used to connect to IoT hub when using the AMQP over web sockets.
```

File: iothub/service/src/ServiceClientTransportSettings.cs [hunk 6]
```diff
@@ -21,13 +21,77 @@ public ServiceClientTransportSettings()
...
+        /// To set a proxy you must instantiate an instance of the <see cref="WebProxy"/> class--or any class that derives from <see cref="IWebProxy"/>. The snippet below shows a method that returns a device using a proxy that connects to localhost on port 8888.
+        /// <code>
+        /// static ServiceClient GetServiceClient()
+        /// {
+        ///     try
+        ///     {
+        ///         var proxyHost = "localhost";
+        ///         var proxyPort = 8888;
+        ///         var proxy = new WebProxy(proxyHost, proxyPort);
+        ///         var transportSettings = new ServiceClientTransportSettings()
```
Comment (drwill-ms) [inline iothub/service/src/ServiceClientTransportSettings.cs:42]:
```suggestion
        ///         var transportSettings = new ServiceClientTransportSettings
```

File: iothub/service/src/ServiceClientTransportSettings.cs [hunk 7]
```diff
@@ -21,13 +21,77 @@ public ServiceClientTransportSettings()
...
+        ///         throw;
+        ///     }
+        /// }
+        /// </code>
+        /// </example>
         public IWebProxy AmqpProxy { get; set; }
 
         /// <summary>
-        /// The proxy settings to be used on the HTTP client.
+        /// The web proxy that will be used to connect to IoT Hub when operations must execute over HTTP.
```
Comment (drwill-ms) [inline iothub/service/src/ServiceClientTransportSettings.cs:61]:
```suggestion
        /// The web proxy that will be used to connect to IoT hub when operations must execute over HTTP.
```

File: iothub/service/src/ServiceClientTransportSettings.cs [hunk 8]
```diff
@@ -21,13 +21,77 @@ public ServiceClientTransportSettings()
...
+        /// To set a proxy you must instantiate an instance of the <see cref="WebProxy"/> class--or any class that derives from <see cref="IWebProxy"/>. The snippet below shows a method that returns a device using a proxy that connects to localhost on port 8888.
+        /// <code>
+        /// static ServiceClient GetServiceClient()
+        /// {
+        ///     try
+        ///     {
+        ///         var proxyHost = "localhost";
+        ///         var proxyPort = 8888;
+        ///         var proxy = new WebProxy(proxyHost, proxyPort);
+        ///         var transportSettings = new ServiceClientTransportSettings()
```
Comment (drwill-ms) [inline iothub/service/src/ServiceClientTransportSettings.cs:79]:
```suggestion
        ///         var transportSettings = new ServiceClientTransportSettings
```

File: iothub/service/src/HttpTransportSettings.cs [hunk 9]
```diff
@@ -20,12 +20,38 @@ public HttpTransportSettings()
...
-        /// Proxy information.
+        /// The web proxy that will be used to connect to IoT Hub when using the HTTP protocol.
         /// </summary>
+        /// <value>
+        /// An instance of a class that implements <see cref="IWebProxy"/>.
+        /// </value>
         /// <remarks>
-        /// This is used when a device is on a network that doesn't have direct internet access and needs to access it via a proxy,
-        /// especially when MQTT and AMQP ports are disallowed to the internet.
+        /// This setting is only valid for the <see cref="JobClient"/> or the <see cref="RegistryManager"/>.
```
Comment (drwill-ms) [inline iothub/service/src/HttpTransportSettings.cs:29]: Worth saying why it'll only work with those clients.

**Review by drwill-ms — APPROVED** [review]

File: iothub/device/src/ITransportSettings.cs (hunk 2 above)
Comment (jamdavi) [reply iothub/device/src/ITransportSettings.cs:45]: Do we have this explicitly defined as a best practice anywhere? I would think that specifying the specific protocol is good for ensuring you choose one over the other, but I think we've just left it up to the customer.

**Pull request merged**